### PR TITLE
fix(yazio): show correct servings in search and barcode scans

### PR DESCRIPTION
**Pull request by thorethy1**
## Description

**What problem does this PR solve?**

This fixes YAZIO servings being displayed incorrectly or incompletely in both the **online food search** and the **barcode scan** flow. It also prevents a selected weighted named serving (for example, `package (400 g)`) from being collapsed after storage or logged under an unrelated default variant when a saved food is deduplicated.

**How did you implement the solution?**

- Group YAZIO search candidates by product ID and merge their top-level servings into the detailed product response.
- Preserve `serving_quantity` for named units while also exposing equivalent `g`/`ml` variants from `amount`.
- Build quantity-aware descriptions and preserve same-name portions with different metric weights.
- Encode metric serving context in persisted named units so `package (200 g)` and `package (400 g)` remain distinct after reload.
- Promote context-rich persisted variants over matching legacy rows when building local picker groups.
- Resolve the selected saved variant only through one exact identity, one genuinely unencoded legacy match, or an exactly matching server default. Ambiguous matches and unrelated defaults now fail closed instead of creating a wrong diary entry.
- Normalize `serving_description: null` to `undefined` at the server schema boundary.
- Add server and mobile regression coverage for search grouping, barcode lookup, persistence, legacy rows, ambiguous matches, and refresh failures.

Linked Issue: #1364

## How to Test

1. From `SparkyFitnessServer/`, run `pnpm run validate` and `SKIP_RLS_MATRIX=1 NODE_ENV=test pnpm run test:ci`.
2. From `SparkyFitnessMobile/`, run `pnpm run validate` and `pnpm exec jest --watchman=false --runInBand`.
3. Search or scan a YAZIO product that reports two same-name weighted servings, such as `package (200 g)` and `package (400 g)`.
4. Verify both options remain visible after saving/reloading and selecting one logs that exact persisted variant.
5. Verify ambiguous legacy/exact matches do not create a diary entry under the default variant.

Local results on the current PR head:

- Server focused schema/YAZIO tests: **30 passed**.
- Server full CI-equivalent suite: **194 files; 2,361 passed, 210 expected RLS skips, 0 failed**.
- Mobile focused serving/persistence tests: **67 passed**.
- Mobile validation: **TypeScript strict + ESLint, 0 warnings**.
- Mobile full suite: **233 suites; 3,837 passed, 0 failed**.
- Independent final review: **no P0/P1/P2 findings**.

## PR Type

- [x] Issue (bug fix)
- [ ] New Feature
- [ ] Refactor
- [ ] Documentation

## Checklist

**All PRs:**

- [x] **[MANDATORY - ALL] Integrity & License**: I certify this is my own work, free of malicious code, and I agree to the [License terms](https://github.com/CodeWithCJ/SparkyFitness/blob/main/LICENSE).

**New features only:**

- [ ] **[MANDATORY for new feature] Alignment**: Not applicable; this is a bug fix related to #1364.

**Frontend changes (`SparkyFitnessFrontend/`):**

- [ ] **[MANDATORY for Frontend changes] Quality**: Not applicable; no frontend files changed.
- [ ] **[MANDATORY for Frontend changes] Translations**: Not applicable; no translations changed.

**Backend changes (`SparkyFitnessServer/`):**

- [x] **[MANDATORY for Backend changes] Code Quality**: Typecheck, lint, formatting, focused tests, and the full CI-equivalent suite pass.
- [x] **[MANDATORY for Backend changes] Database Security**: No tables, migrations, SQL, or RLS policies changed.

**UI changes (components, screens, pages):**

- [ ] **[MANDATORY for UI changes] Screenshots**: Not applicable; serving identity and save resolution changed without visual layout changes.

**Mobile changes (`SparkyFitnessMobile/`):**

- [x] **[MANDATORY for Mobile changes] Tested on device or emulator**: Tested on Device via Development Build on iPhone 16 pro Max.

## Screenshots

Not applicable; no visual layout changed.

## Notes for Reviewers

- Production behavior changes span the YAZIO server mapping and the mobile serving persistence/selection path.
- No database migration, lockfile update, new endpoint, frontend change, or shared-schema change is included.
- Distinct metric contexts remain separate even when an older undecorated serving row loads first.
- Entry creation intentionally aborts when the selected persisted variant cannot be resolved uniquely.

## Summary by CodeRabbit

- **New Features**
  - Added richer YAZIO serving handling (named servings and gram-weight equivalents), with serving descriptions preserved during searches and barcode scans.
  - Improved merging of related YAZIO results into complete variant sets for food entries.
- **Bug Fixes**
  - Fixed persisted serving/variant resolution with clearer behavior for ambiguous or refresh-failing scenarios.
  - Prevented distinct serving weights from being incorrectly merged.
  - Normalized `null` serving descriptions consistently in API payloads and validation.
- **Tests**
  - Expanded automated coverage for YAZIO serving mapping, persistence, and variant resolution edge cases.
- **Documentation**
  - Updated Jest run guidance for watchman-disabled testing.

### CODE REVIEW

File: SparkyFitnessMobile/__tests__/utils/persistExternalVariants.test.ts [hunk 1]
```diff
@@ -71,6 +71,75 @@ describe('persistExternalVariants', () => {
...
+    expect(mockedCreateFoodVariant).toHaveBeenCalledWith(
+      expect.objectContaining({
+        food_id: 'food-1',
+        serving_size: 200,
+        serving_unit: 'g',
+        calories: 100,
+      }),
+    );
+  });
+
```
[the remark this answers is not in the document]
Comment (thorethy1) [reply SparkyFitnessMobile/__tests__/utils/persistExternalVariants.test.ts]: Fixed in 33020547 without a database migration. Named servings now preserve distinguishing metric context in the existing persisted serving_unit (for example, `package (200 g)` and `package (400 g)`), while metric units remain unchanged and unambiguous legacy rows still work. Selection and diary-entry resolution use the persisted identity so the correct variant ID/unit survives import. Regression coverage includes both package weights, external search, persistence, legacy fallback, and entry creation. Verified with Mobile validate plus the full suite: 233 suites / 3830 tests.

**Comment by thorethy1** [comment]
### Related PR: #1924

PR #1924 is complementary to this fix, not overlapping:

- **#1933** preserves and resolves YAZIO serving variants such as `package (200 g)` vs. `package (400 g)` across search, barcode lookup, persistence, and mobile selection.
- **#1924** normalizes AI `create_food` units such as `4 pieces` and enforces a positive `serving_size` invariant at the database layer.

The PRs share no changed files, a clean merge simulation produced no conflicts, and the YAZIO mapping here already rejects non-positive amounts / falls back to a positive serving quantity. They can therefore be reviewed and merged independently; #1924 would provide an additional database-level safeguard.

**Comment by CodeWithCJ** [comment]
THanks for your PR. Could you share example foods to search to test this PR. and also attach some screenshots.

**Comment by CodeWithCJ** [comment]
I am seeing tick mark only on Yazio result. why is that? 

image

File: SparkyFitnessMobile/AGENTS.md [hunk 2]
```diff
@@ -33,7 +33,7 @@ pnpm run android
 pnpm run lint
 pnpm run typecheck
 pnpm run validate
-pnpm run test:run -- --watchman=false --runInBand
+pnpm exec jest --watchman=false --runInBand
```
Comment (CodeWithCJ) [inline SparkyFitnessMobile/AGENTS.md:36]: I don't think this change is required.

**Comment by CodeWithCJ** [comment]
@apedley  could you review this

**Comment by thorethy1** [comment]
> I am seeing tick mark only on Yazio result. why is that? 
> 
> 
> 
> image
> 
> 

Do you mean you only get results with checkmarks? 

Yazio is a DB powered by the community. Everyone can create foods there. The foods with checkmarks are checked by Yazio to have the right nutrition. 
Most generic foods (like chicken or apples) have those checkmarks, but many foods you scan by barcode don’t.

If you search for something specific, it is much more likely to find something non-verified.

**Comment by thorethy1** [comment]
Thanks for the review — below are reproducible YAZIO search examples and a visual before/after comparison.

## Search examples

### Verified
Search for these exact terms:

- `Apfel` (Apple)
- `Banane` (Banana)
- `Orange`
- `Hühnerei, Eier, gekocht` (boiled eggs)
- `Apfelmus (ungesüßt)` (unsweetened applesauce)
- `Joghurt 1,5% Fett` (1.5% fat yogurt)

### Additional examples to verify

- `Fettarmer Naturjoghurt Milbona` (Milbona low-fat natural yogurt)
- `oil spray esn`
- `Monster Ultra`
- `Chiasamen Rewe Bio`
- `Laugenbrezel Ditsch`

## Before

### YAZIO barcode lookup
YAZIO barcode lookup before

### YAZIO online search
YAZIO online search before

### Food saved locally from YAZIO
Locally saved YAZIO food before

## After

### YAZIO online search
YAZIO online search after

### Food saved locally from YAZIO
Locally saved YAZIO food after
Locally saved YAZIO food after — serving selection

### YAZIO barcode lookup
YAZIO barcode lookup after

**Comment by apedley** [comment]
> I am seeing tick mark only on Yazio result. why is that?

@CodeWithCJ The check marks appear for foods that are verified by yazio and the badge was added to web and mobile around 2 weeks ago.

File: SparkyFitnessServer/integrations/yazio/yazioService.ts [hunk 3]
```diff
@@ -676,30 +787,40 @@ async function searchYazioFoods(query: string, options: YazioSearchOptions) {
...
   const offset = Math.max(page - 1, 0) * pageSize;
-  const pageItems = products.slice(offset, offset + pageSize);
+  const pageGroups = productGroups.slice(offset, offset + pageSize);
 
   const foods = await Promise.all(
-    pageItems.map(async (product) => {
-      const productId = product.id ?? product.product_id;
+    pageGroups.map(async ({ product, candidates: siblingCandidates }) => {
+      const productId = getYazioProductId(product);
       if (!productId) {
```
Comment (apedley) [inline SparkyFitnessServer/integrations/yazio/yazioService.ts]: This whole code branch is dead now. groupYazioSearchCandidates skips products with no id so every group has one

File: SparkyFitnessMobile/src/hooks/useAddFoodEntry.ts
Comment (apedley) [inline SparkyFitnessMobile/src/hooks/useAddFoodEntry.ts]: This says "please try again" but trying again doesn't do anything anymore

**Review by apedley — COMMENTED** [review]
This looks really good. Left two minor cleanup comments. Really appreciate that you included tests!

**Comment by thorethy1** [comment]
> This looks really good. Left two minor cleanup comments. Really appreciate that you included tests!

Thanks for reviewing! I can imagine it sometimes takes you longer to review than doing it yourself. Will clean those two things up.

File: SparkyFitnessServer/integrations/yazio/yazioService.ts (hunk 3 above)
Comment (thorethy1) [reply SparkyFitnessServer/integrations/yazio/yazioService.ts]: Fixed in 7cbd9320. `groupYazioSearchCandidates` now carries the already-validated `productId` into each group, so the unreachable fallback is removed while keeping the ID non-optional at the enrichment call site. Targeted YAZIO tests and server validation pass.

File: SparkyFitnessMobile/src/hooks/useAddFoodEntry.ts
Comment (thorethy1) [reply SparkyFitnessMobile/src/hooks/useAddFoodEntry.ts]: Fixed in 7cbd9320. The fail-closed selected-serving resolution now tells the user `Choose a different serving.` rather than suggesting a retry that cannot resolve an ambiguous/missing serving. Added a regression assertion; focused mobile tests and mobile validation pass.

**Comment by apedley** [comment]
@CodeWithCJ looks good to me!

**Review by CodeWithCJ — APPROVED** [review]
